src/main/java/steps/CartPageSteps.java

```diff
@@ -16,4 +23,58 @@ public static void goToCheckout() {
         cartPage = new CartPage();
         cartPage.clickOnFinalCheckoutButton();
     }
-}
+
+    public static List<ProductModel> getProductsFromShoppingCartTable() {
+        cartPage = new CartPage();
+        return cartPage.shoppingCartTable.getProducts();
+    }
+
+    public static List<CheckoutProductModel> getCheckoutProductsFromShoppingCartTable() {
+        cartPage = new CartPage();
+        return cartPage.shoppingCartTable.getCheckoutProducts();
+    }
+
+    public static void removeProductFromShoppingCartTable(CheckoutProductModel checkoutProduct) {
+        cartPage = new CartPage();
+        var shoppingCartTableRow = new ShoppingCartTableRow(cartPage.shoppingCartTable
+                .getCheckoutProducts()
+                .indexOf(checkoutProduct) + 1);
+        shoppingCartTableRow.removeProduct();
+    }
+
+    public static void changeQuantityOfProduct(ProductModel product, int quantity) {
+        cartPage = new CartPage();
+        var shoppingCartTableRow = new ShoppingCartTableRow(cartPage.shoppingCartTable
+                .getProducts()
+                .indexOf(product) + 1);
+        shoppingCartTableRow.setQuantity(quantity);
+    }
+
+    public static void checkIfCartIsEmpty() {
+        cartPage = new CartPage();
+        Assert.assertTrue(cartPage.shoppingCartTable.isCartEmpty());
+    }
+
+    public static void checkIfProductAppearsOnce(ProductModel product, List<CheckoutProductModel> checkoutProducts) {
+        var count = checkoutProducts.stream()
+                .filter(p -> p.getProductModel().equals(product))
+                .count();
+        Assert.assertEquals(1, count);
+    }
+
+    public static void checkIfTotalPriceIsCorrect(List<ProductModel> products, List<CheckoutProductModel> checkoutProducts) {
+        var unitPrices = products.stream()
+                .map(ProductModel::getPrice)
+                .toList();
+        var quantities = checkoutProducts.stream()
+                .map(CheckoutProductModel::getQuantity)
+                .toList();
+        var totalPrices = checkoutProducts.stream()
```

Comment:
Maybe something like:

 checkoutProducts.stream()
                .forEach(product -> assertThat(product.getTotalPrice().isEqual(product.quantity * getProductByCheckoutProduct(product)))
                
               where getProductByCheckoutProduct() returns you single product from products List which matches the name of the CheckoutProductModel?

Reply:
Rozdzieliłem te metodę na dwie, dzięki
```java
    public static void checkIfTotalPricesAreCorrectPerProduct(List<ProductModel> products, List<CheckoutProductModel> checkoutProducts) {
        checkoutProducts.forEach(checkoutProduct -> {
            ProductModel product = getProductByCheckoutProduct(products, checkoutProduct);
            BigDecimal expectedTotalPrice = product.getPrice().multiply(BigDecimal.valueOf(checkoutProduct.getQuantity()));
            Assert.assertEquals(expectedTotalPrice, checkoutProduct.getTotalPrice());
        });
    }

    public static ProductModel getProductByCheckoutProduct(List<ProductModel> products, CheckoutProductModel checkoutProduct) {
        return products.stream()
                .filter(product -> product.getName().equals(checkoutProduct.getProductModel().getName()))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("No matching product found for checkout product"));
    }
```

Reply:
I split the method into two methods as you wrote
```java
    public static void checkIfTotalPricesAreCorrectPerProduct(List<ProductModel> products, List<CheckoutProductModel> checkoutProducts) {
        checkoutProducts.forEach(checkoutProduct -> {
            ProductModel product = getProductByCheckoutProduct(products, checkoutProduct);
            BigDecimal expectedTotalPrice = product.getPrice().multiply(BigDecimal.valueOf(checkoutProduct.getQuantity()));
            Assert.assertEquals(expectedTotalPrice, checkoutProduct.getTotalPrice());
        });
    }

    public static ProductModel getProductByCheckoutProduct(List<ProductModel> products, CheckoutProductModel checkoutProduct) {
        return products.stream()
                .filter(product -> product.getName().equals(checkoutProduct.getProductModel().getName()))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("No matching product found for checkout product"));
    }
```